pkg/identityserver/application_registry.go

```diff
@@ -83,6 +83,7 @@ func (is *IdentityServer) getApplication(ctx context.Context, req *ttnpb.GetAppl
 	if err = is.RequireAuthenticated(ctx); err != nil {
 		return nil, err
 	}
+	req.FieldMask.Paths = cleanFieldMaskPaths(ttnpb.ApplicationFieldPathsNested, req.FieldMask.Paths, getPaths, nil)
```

Comment:
Why is this necessary?
Why not just set this to `getPaths` if `len(req.FieldMask.Paths) == 0`?
Basically the same question for all other invocations - I really don't get what's going on here.

Reply:
`getPaths` contains the fields that are implicitly part of the field mask, so they are always added to the field mask in the request, if it doesn't already contain them.

Reply:
Yeah, but there's no harm in adding a field twice - you can just append the fields - we do the same in JS/NS and AS AFAIK.

Reply:
In order to construct the slice of required fields, `Set` registry endpoint in NS constructs a `gets` and `sets`: https://github.com/TheThingsNetwork/lorawan-stack/blob/59154c3b694652b4c72447ae4f74c606003dac80/pkg/networkserver/grpc_deviceregistry.go#L49-L57

Reply:
This is not the NS/JS. In the IS we prefer to have paths in the field mask only once.

Reply:
But then why not `ttnpb.EnsureFieldpaths(paths []string, ensure ...string) []string`?
You pass the existing fieldmask and the paths you want to add if missing.
You don't need 4 arguments for that.

Reply:
There is no `ttnpb.EnsureFieldPaths`, I want to combine this with the allowed paths, I want to be able to specify paths that need to be deleted if present, and finally, I want to do it in a single pass instead of combining `EnsureFieldPaths`, `EnsureNotFieldPaths`, `IntersectFieldPaths`, `UniqueFieldPaths`.

Reply:
1. We have API layer middleware, which should validate that fields that should not be present are not present.
2. Invalid fieldpaths, which still could get through should be validated explicitly and if present - throw an error.

RPC implementation simply should not delete anything passed to it. Cleaning and validation belongs to the API layer middleware. RPC should simply upsert the required fields.